src/index.ts

```diff
@@ -854,6 +870,59 @@ export abstract class PinejsClientCore<PinejsClient> {
 		}
 	}
 
+	private canRetryDefaultHandler(err: any) {
+		const code = err?.statusCode;
+		return !code || code === 429 || (code >= 500 && code < 600);
+	}
+
+	protected async callWithRetry<T>(
+		fnCall: () => Promise<T>,
+		retry: RetryParameters,
+	): Promise<T> {
+		if (!retry) {
+			return await fnCall();
+		}
+
+		const defaultRetry = this.retry || {};
+		const minDelayMs = retry.minDelayMs ?? defaultRetry.minDelayMs ?? 0;
+		const maxDelayMs = retry.maxDelayMs ?? defaultRetry.maxDelayMs ?? 0;
+		const maxAttempts = retry.maxAttempts ?? defaultRetry.maxAttempts ?? 0;
+
+		if (
+			minDelayMs <= 0 ||
+			maxDelayMs <= 0 ||
+			maxAttempts <= 0 ||
+			minDelayMs > maxDelayMs
+		) {
+			console.warn('pinejs-client invalid retry parameters');
+			return await fnCall();
+		}
+
+		const onRetryHandler = retry.onRetry ?? defaultRetry.onRetry;
+
+		let attempt = 0;
+
+		const canRetryHandler =
+			retry.canRetry ?? defaultRetry.canRetry ?? this.canRetryDefaultHandler;
+
+		while (true) {
+			try {
+				return await fnCall();
+			} catch (err) {
+				if (attempt < maxAttempts && !canRetryHandler(err)) {
```

Comment:
I think you want this to be
```suggestion
				if (attempt >= maxAttempts || !canRetryHandler(err)) {
```
since right now it'll only throw/exit the loop if it's *both* within the expected retry count and the `canRetryHandler` also returns false, whereas what we want is either it to be out of attempt or not able to retry

Reply:
Yeah, I just saw that... also, we have a small problem here in that we will retry one too many times.  I need to fiddle with `attempt`

Reply:
Yeah I think it might make sense to start at `attempt = 1` and do `(attempt - 1)` below instead of `attempt + 1`

Reply:
I was going to do `if (attempt >= (maxAttempts - 1) && !canRetryHandler(err))` but that works too.